# **digitalDLSorteR** <img src="man/figures/logo.png" align="right" width="120"/>

<div style="text-align:left">
<span>
<h4>An R package to deconvolute bulk RNA-Seq from scRNA-Seq data based on Deep Learning</h4></span>
</div>

<br>


The **digitalDLSorteR** R package provides a set of tools to deconvolve and infer cell type proportions of bulk RNA-seq data through the development of context-specific deconvolution models based on Deep Learning and single-cell RNA-Seq (scRNA-Seq) data. These models are able to accurately enumerate and quantify cell proportions of bulk RNA-Seq data from specific biological environments. For more details about the algorithm and the functionalities implemented in this package, see [Torroja and Sanchez-Cabo, 2019](https://www.frontiersin.org/articles/10.3389/fgene.2019.00978/full) and `vignette(digitalDLSorteR)`.

## Installation

**digitalDLSorteR** can be installed from GitHub, although it will be available at CRAN coming soon.

```r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("diegommcc/digitalDLSorteR")
```

**digitalDLSorteR** depends on **digitalDLSorteRdata** data package as it makes available pre-trained context-specific models deconvolution and other data needed for tests and examples. Therefore, it must be installed together with **digitalDLSorteR** as one of its dependencies.

## Rationale of **digitalDLSorteR**

This algorithm consists of training Deep Neural Network (DNN) models with simulated bulk RNA-Seq samples whose cell composition is known. These 'pseudo-bulk RNA-Seq' samples are generated by aggregating pre-characterized scRNA-Seq data from specific biological environments. These models are able to accurately deconvolve new bulk RNA-Seq samples from the same environment since they are able to consider possible environmental-dependent transcriptional changes of specific cells, such as macrophages or other immune cells in complex diseases (e.g., specific subtypes of cancer or atherosclerosis). This aspect allows to overcome this limitation presents in other methods. For instance, in the case of immune cells, published methods usually rely on purified transcriptional profiles of peripheral blood mononuclear cells despite the fact these cells are highly variable depending on the environment where they are. Therefore, considering this feature together with the use of powerful DNN models and the fact that improvement of scRNA-Seq datasets over time will allow to build better models, **digitalDLSorteR** offers a good alternative to classical methods based on linear models with pre-defined markers and irreal transcriptional references.

## Usage

The package has two main ways of usage:

1. Using pre-trained models included in `digitalLDSorteRdata` package to deconvolve new bulk RNA-Seq samples from the same environment. So far, the available models allow to deconvolve samples from human breast cancer ([GSE75688](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688) from [Chung et al., 2017](https://www.nature.com/articles/ncomms15081) used as reference), and colorectal cancer ([GSE81861](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81861) from [Li et al., 2017](https://www.nature.com/articles/ng.3818) used as reference). For more details about this workflow, please see `vignette(digitalDLSorteR)` or . Moreover, in the panel A of the figure below it is summarized.
2. Building new deconvolution models from pre-characterized scRNA-Seq datasets. This workflow involves some requirements regarding computational resources, although **digitalDLSorteR** has been developed to provide with all the possible facilities to make the process easier, such as batch processing of data and the use of the `HDF5Array` and `DelayedArray` packages. For more information about this workflow, see `vignette(digitalDLSorteR)`. Moreover, the main steps are summarized in the panel B of the figure below.

<img src="man/figures/workflow_readme.png"/>

## Final remarks

* Regarding available pre-trained models, new models coming soon! Of course, you can build your own models and even contribute to **digitalDLSorteR** making them available. Contact us to incude them in the package!
* Report bugs at <https://github.com/diegommcc/digitalDLSorteR/issues>
* Contributions and suggestions are welcome!

## References

<table>
  <tr><td>Chung, W., Eum, H. H., Lee, H. O., Lee, K. M., Lee, H. B., Kim, K. T., et al. (2017). Single-cell RNA-seq enables comprehensive tumour and immune cell profiling in primary breast cancer.
  <i>Nat. Commun.</i>
  <b>8</b> (1) 15081
  <a href='https://doi.org/10.1038/ncomms15081'>doi:10.1038/ncomms15081</a>
  </td></tr>

  <tr><td>Li, H., Courtois, E. T., Sengupta, D., Tan, Y., Chen, K. H., Goh, J. J. L., et al. (2017). Reference component analysis of single-cell transcriptomes elucidates cellular heterogeneity in human colorectal tumors.
  <i>Nat. Genet.</i>
  <b>49</b> (5), 708-718
  <a href='https://doi.org/10.1038/ng.3818'>doi:10.1038/ng.3818</a>
  </td></tr>

  <tr><td>Torroja, C. and SÃ¡nchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning algorithm to quantify immune cell populations based on scRNA-Seq data.
  <i>Frontiers in Genetics</i>
  <b>10</b> 978
  <a href='https://doi.org/10.3389/fgene.2019.00978'>doi:10.3389/fgene.2019.00978</a>
  </td></tr>
</table>
